src/data_chain/vote_accumulator.rs

```diff
@@ -39,33 +38,136 @@ pub enum AccumulationReturn<T> {
     ValidProof(Proof), // block corresponding to this has already accumulated - add to the chain
 }
 
-#[derive(Debug, Default)]
+#[derive(Default)]
 pub struct VoteAccumulator<T: Ord> {
-    votes: BTreeMap<T, (Vec<Proof>, Instant)>,
+    blocks: BTreeMap<T, State<T>>,
 }
 
-impl<T: Ord> VoteAccumulator<T> {
+impl<T: Clone + Ord + Serialize> VoteAccumulator<T> {
     // FIXME - remove these
     #[allow(unused)]
-    #[cfg_attr(feature = "cargo-clippy", allow(needless_pass_by_value))]
-    pub fn add_vote<'a, I: IntoIterator<Item = &'a PublicInfo>>(
+    pub fn add_vote(
         &mut self,
-        _vote: Vote<T>,
-        _node_info: &PublicInfo,
-        _valid_nodes_itr: I,
+        vote: &Vote<T>,
+        node_info: &PublicInfo,
+        valid_nodes: &BTreeSet<PublicInfo>,
     ) -> Result<Vec<AccumulationReturn<T>>, RoutingError> {
-        Ok(vec![])
+        let mut results = self.expire(valid_nodes);
+
+        match self.blocks.entry(vote.payload().clone()) {
+            Entry::Vacant(entry) => {
+                let _ = entry.insert(State {
+                    block: Block::new(vote, node_info)?,
+                    timestamp: Instant::now(),
+                });
+            }
+            Entry::Occupied(mut entry) => {
+                let valid_before = entry
+                    .get()
+                    .block
+                    .get_block_state(valid_nodes)
+                    .valid_or_full();
+                let valid_after = entry
+                    .get_mut()
+                    .block
+                    .add_vote(vote, node_info, valid_nodes)?
+                    .valid_or_full();
+
+                if valid_after {
```

Comment:
looks like we don't need `valid_after` variable ?

Reply:
We don't necessarily, but I think it makes the code slightly more readable, so I'd prefer to keep it as is.